### PR TITLE
bugfix #42 non-ascii redirect fix url and query params encodings

### PR DESCRIPTION
**Pull request by ovidiul**
Fixed the non-ASCII redirect issue.

The main issue is the wp_parse_url function which breaks the encoding, and that was replaced with a method to support that.

Also fixes the $_SERVER and QUERY_STRING returned by decoding it.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,15 +6,18 @@ notifications:
     on_failure: change
 
 php:
-  - 5.6
```
Comment (GaryJones) [inline .travis.yml]: The plugin may be being used outside of VIP, so a change in minimum PHP would be a breaking change, and therefore require a major version bump. For now, let's keep 5.6 as the minimum and have that as a separate discussion.

File: .travis.yml [hunk 2]
```diff
@@ -6,15 +6,18 @@ notifications:
...
-  - 7.0
-  - 7.1
+  - 7.3
+  - 7.4
+
+services:
+  - mysql
```
Comment (GaryJones) [inline .travis.yml]: These are good fixes, but they should be moved into their own PR please.

File: composer.json [hunk 3]
```diff
@@ -1,5 +1,4 @@
-{
-  "name": "automattic/wpcom-legacy-redirector",
+{                                                                                                                                                                                                                              "name": "automattic/wpcom-legacy-redirector",
```
Comment (GaryJones) [inline composer.json]: What's with the long whitespace here?

File: composer.json [hunk 4]
```diff
@@ -16,5 +15,11 @@
   "support": {
     "issues": "https://github.com/Automattic/WPCOM-Legacy-Redirector/issues",
     "source": "https://github.com/Automattic/WPCOM-Legacy-Redirector"
+  },
+  "require-dev": {
```
Comment (GaryJones) [inline composer.json]: Again, good to get these dependencies added, but should be handled in its own PR.

File: composer.json [hunk 5]
```diff
@@ -16,5 +15,11 @@
...
+    "automattic/vipwpcs": "^2.2",
+    "dealerdirect/phpcodesniffer-composer-installer": "^0.7.1",
```
Comment (GaryJones) [inline composer.json]: We can make the constraint here much wider: `^0.4.1 || ^0.5 || ^0.6.2 || ^0.7`.

File: composer.json [hunk 6]
```diff
@@ -16,5 +15,11 @@
...
+    "phpunit/phpunit": "^7"
```
Comment (GaryJones) [inline composer.json]: Likewise: `^4 || ^5 || ^6 || ^7` gives more flexibility about which version is installed. Until there is something just from PHPUnit 7 which is being used (and can then likely use https://github.com/Yoast/PHPUnit-Polyfills anyway to make the PHPUnit 7 thing available in earlier versions anyway), then keep the constraint as loose as possible (then supports PHP 5.6 as well).

File: composer.json [hunk 7]
```diff
@@ -16,5 +15,11 @@
...
   }
 }
+
```
Comment (GaryJones) [inline composer.json]: Don't include the `composer.lock` file in this package, as that prescribes what versions of dependencies are working for just your set-up - someone with a different PHP version may need other versions of dependencies.

File: includes/class-wpcom-legacy-redirector.php [hunk 8]
```diff
@@ -3,513 +3,570 @@
...
-			} elseif ( 'publish' !== get_post_status( $parent ) ) {
-				return 'private';
-			} else {
-				$parent_slug = $parent->post_name;
-				return $parent_slug;
-			}
-		}
-	}
+class WPCOM_Legacy_Redirector
+{
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php:7]: This isn't standard position a brace in the WordPress community.

File: includes/class-wpcom-legacy-redirector-cli.php [hunk 9]
```diff
@@ -81,7 +81,11 @@ function insert_redirect( $args, $assoc_args ) {
 		$inserted = WPCOM_Legacy_Redirector::insert_legacy_redirect( $from_url, $to_url );
 
 		if ( ! $inserted || is_wp_error( $inserted ) ) {
-			WP_CLI::error( sprintf( "Couldn't insert %s -> %s", $from_url, $to_url ) );
+			$error_text = "";
+			if (  is_wp_error( $inserted ) ) {
+				$error_text =  $inserted->get_error_message();
+			}
+			WP_CLI::error( sprintf( "Couldn't insert %s -> %s (%s)", $from_url, $to_url, $error_text ) );
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector-cli.php:88]: This looks like it's a good fix to go into it's own PR (ideally with tests).

File: includes/class-wpcom-legacy-redirector.php [hunk 10]
```diff
@@ -3,513 +3,570 @@
...
+     * @return string Hash representation of string.
+     */
+    private static function get_url_hash( $url )
+    {
+        return md5($url);
+    }
+
+    /**
+     * UTF-8 aware parse_url() replacement.
+     *
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php:343]: Needs some `@param` entries.

File: includes/class-wpcom-legacy-redirector.php [hunk 11]
```diff
@@ -3,513 +3,570 @@
...
+    {
+        return md5($url);
+    }
+
+    /**
+     * UTF-8 aware parse_url() replacement.
+     * 
+     * @return array
+     */
+    public static function mb_parse_url($url, $component = -1)
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php:346]: This whole method should be written using the WP coding style - lots of whitespace, particularly inside `( )`.

**Review by GaryJones — CHANGES_REQUESTED** [review]
The main fix looks good - but let's keep changes in a PR down to a minimum, spin up new PRs for unrelated changes to the project, and follow the WordPress (and VIP) coding style / standards.

File: .travis.yml (hunk 1 above)
Comment (ovidiul) [reply .travis.yml]: rebased the PR to include this change from fix/travix

File: .travis.yml (hunk 2 above)
Comment (ovidiul) [reply .travis.yml]: moved to fix/travis per your contribution!

File: composer.json (hunk 3 above)
Comment (ovidiul) [reply composer.json]: fixed!

File: composer.json (hunk 5 above)
Comment (ovidiul) [reply composer.json]: done in fix/travis branch

**Comment by GaryJones** [comment]
Closing in favour of #92.
